### PR TITLE
Changes to message serialization, signing and verification

### PR DESCRIPTION
**Pull request by mfranciszkiewicz**
- signature verification does not require message re-serialization
- signature verification is no longer affected by the order of properties inside objects

Addresses #1457

### CODE REVIEW

File: golem/network/transport/message.py [hunk 1]
```diff
@@ -11,123 +11,117 @@
...
-                encrypted = False
-            except Exception as err:
-                logger.info("Failed to decrypt message {}".format(str(err)))
-                continue
+        11 bytes in total
 
-            m = cls.deserialize_message(msg)
+        :return: serialized header
+        """
+        return struct.pack('!HQ?', self.TYPE,
```
Comment (etam) [inline golem/network/transport/message.py]: E: golem/network/transport/message.py:102:35: Instance of 'Message' has no 'TYPE' member (no-member)

File: golem/core/databuffer.py [hunk 2]
```diff
@@ -21,19 +18,12 @@ def append_ulong(self, num):
...
-        :param check_size: keep buffer size below MAX_BUFFER_SIZE
-        :param overflow_prefix: string to prepend on overflow
         :param str data: string to append
         """
-        new_size = self.data_size() + len(data)
-        if check_size and new_size > MAX_BUFFER_SIZE:
-            self.buffered_data = b"".join([overflow_prefix or '', data])
-        else:
-            self.buffered_data = b"".join([self.buffered_data, data])
+        self.buffered_data = b"".join([self.buffered_data, data])
```
Comment (etam) [inline golem/core/databuffer.py]: Is this bytes or strings?

```python
>>> b''.join([b'123', "abc"])
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: sequence item 1: expected a bytes-like object, str found
```

If it's all bytes, then this should be fine:
```python
self.buffered_data += data
```

File: golem/core/databuffer.py [hunk 3]
```diff
@@ -1,9 +1,6 @@
 import struct
-from golem.core.common import to_unicode
 
-from .variables import LONG_STANDARD_SIZE
-
-MAX_BUFFER_SIZE = 2 * 1024 * 1024
+from .variables import LONG_STANDARD_SIZE
 
 
 class DataBuffer:
```
Comment (etam) [inline golem/core/databuffer.py]: Comments in this class suggests, that it operates on strings, but instead it operates on bytes.

File: golem/core/variables.py [hunk 4]
```diff
@@ -5,7 +5,7 @@
 #       VARIABLES          #
 ############################
 APP_NAME = "Brass Golem"
-APP_VERSION = "0.8.0"
```
Comment (etam) [inline golem/core/variables.py:8]: Is this supposed to go in here?

File: golem/network/transport/message.py [hunk 5]
```diff
@@ -11,123 +11,117 @@
...
-        :param SafeServer server: server that is able to decrypt data
-        :return list: list of decrypted and deserialized messages
-        """
-        if not isinstance(db_, DataBuffer):
-            raise TypeError(
-                "Incorrect db type: {}. Should be: DataBuffer".format(db_)
+            if sign_func:
+                self.sig = sign_func(self.get_short_hash())
+            else:
+                self.sig = b'0' * 65
```
Comment (etam) [inline golem/network/transport/message.py]: Magic constant

File: tests/golem/network/transport/test_network.py [hunk 6]
```diff
@@ -27,8 +27,7 @@ def interpret(self, msg):
         self.msgs.append(msg)
 
     def sign(self, msg):
-        msg.sig = "ASessionSign"
-        return msg
+        return b'1' * 65
```
Comment (etam) [inline tests/golem/network/transport/test_network.py]: Magic constant

**Review by etam — CHANGES_REQUESTED** [review]

File: golem/network/transport/message.py [hunk 7]
```diff
@@ -11,123 +11,117 @@
...
         self.sig = sig
-        if timestamp is None:
-            timestamp = time.time()
-        self.timestamp = timestamp
-        self.encrypted = False  # inform if message was encrypted
 
-        self.load_dict_repr(dict_repr)
+        # Encoded data
+        self._payload = payload  # child's payload only (may be encrypted)
+        self._raw = raw  # whole message
```
Comment (Krigpl) [inline golem/network/transport/message.py]: I might be missing something but is it used at all?

File: golem/network/transport/message.py [hunk 8]
```diff
@@ -11,123 +11,117 @@
...
+        self.encrypted = encrypted
         self.sig = sig
-        if timestamp is None:
-            timestamp = time.time()
-        self.timestamp = timestamp
-        self.encrypted = False  # inform if message was encrypted
 
-        self.load_dict_repr(dict_repr)
+        # Encoded data
+        self._payload = payload  # child's payload only (may be encrypted)
```
Comment (Krigpl) [inline golem/network/transport/message.py:51]: `self.encrypted`, `self.sig` and `self._payload` don't need to exist.
They are only set and used during serialization, we can just used them there locally.

File: golem/network/transport/message.py (hunk 7 above)
Comment (Krigpl) [inline golem/network/transport/message.py:52]: I might be missing something but is it used?

File: golem/core/databuffer.py (hunk 3 above)
Comment (mfranciszkiewicz) [reply golem/core/databuffer.py]: Comments weren't changed while porting from Python 2 to Python 3.

File: golem/core/databuffer.py (hunk 2 above)
Comment (mfranciszkiewicz) [reply golem/core/databuffer.py]: Yes, it should be fine. `''.join` is just faster.

File: golem/network/transport/message.py [hunk 9]
```diff
@@ -11,123 +11,117 @@
...
+        self._raw = raw  # whole message
+
+    @property
+    def raw(self):
+        """Returns a raw copy of the message"""
+        return self._raw[:]
 
     def get_short_hash(self):
```
Comment (Krigpl) [inline golem/network/transport/message.py:59]: I believe this doesn't have to be exposed anymore.
And by making it an internal method we can get rid of the properties I mentioned in the other comment

File: golem/network/transport/message.py (hunk 5 above)
Comment (mfranciszkiewicz) [reply golem/network/transport/message.py]: Actually I've tried replacing all byte counts with constants and the code was IMO less readable. I'd like to hear another opinion.

File: golem/core/variables.py (hunk 4 above)
Comment (mfranciszkiewicz) [reply golem/core/variables.py:8]: The setup script automatically updates the version and it's pointless to fight it with every PR.

File: tests/golem/network/transport/test_network.py (hunk 6 above)
Comment (mfranciszkiewicz) [reply tests/golem/network/transport/test_network.py]: I believe that this is ok in tests.

File: golem/network/transport/message.py (hunk 1 above)
Comment (mfranciszkiewicz) [reply golem/network/transport/message.py]: The base `Message` class is not instantiated in the application. Where is that message from?

**Review by Krigpl — CHANGES_REQUESTED** [review]
Overall looks good, just a couple of comments to make the Message class even a bit simpler.

File: golem/network/transport/message.py (hunk 7 above)
Comment (mfranciszkiewicz) [reply golem/network/transport/message.py:52]: This will be used in the near future

File: golem/network/transport/message.py (hunk 9 above)
Comment (mfranciszkiewicz) [reply golem/network/transport/message.py:59]: I agree on hiding the method. EDIT: actually, it's still needed outside the class.

File: golem/core/databuffer.py (hunk 2 above)
Comment (etam) [reply golem/core/databuffer.py]:
```
$ python3 -m timeit 'b1 = b"0"*100; b2 = b"f"*100; b1 = b"".join([b1, b2])'
10000000 loops, best of 3: 0.19 usec per loop
$ python3 -m timeit 'b1 = b"0"*100; b2 = b"f"*100; b1 += b2'
10000000 loops, best of 3: 0.126 usec per loop
```
It looks like `+=` is faster.

File: golem/network/transport/message.py (hunk 1 above)
Comment (etam) [reply golem/network/transport/message.py]: From pylint, run by shippable.

File: golem/core/databuffer.py (hunk 2 above)
Comment (mfranciszkiewicz) [reply golem/core/databuffer.py]: Though 100 bytes is by orders of magnitude too low for this test, it seems that `b''.join` is not faster in Python 3.

Comment (mfranciszkiewicz) [reply golem/core/databuffer.py]: I'll revert changes all changes made in databuffer and address them in a separate PR.

Comment (etam) [reply golem/core/databuffer.py]: At 10000 bytes they become equal

File: golem/network/transport/message.py (hunk 8 above)
Comment (mfranciszkiewicz) [reply golem/network/transport/message.py:51]: Without more refactoring, these are still needed for signature verification, done outside the class.

File: golem/network/transport/message.py (hunk 5 above)
Comment (Krigpl) [reply golem/network/transport/message.py]: It's used three times and in one place you put a comment explaining what it is so might be worth using an actual constant.
Something like `SIG_LEN = 65` shouldn't hurt legibility too much

File: golem/network/transport/message.py [hunk 10]
```diff
@@ -1,8 +1,8 @@
-import collections
 import logging
+import struct
 import time
 
-from golem.core.common import to_unicode
+
 from golem.core.databuffer import DataBuffer
```
Comment (Krigpl) [inline golem/network/transport/message.py]: I think it'd be nice if we'd get rid of DataBuffer here. That would mean removing `deserialize` method and leaving only `deserialize_message` which doesn't depend on DataBuffer. `deserialize` is only used in `tcpnetwork.py` and that's where DataBuffer logic should live. You already did pretty much the same with `decrypt_and_deserialize` so removing DataBuffer dependency altogether would be nice.

**Review by Krigpl — APPROVED** [review]
I'm fine with the current version, just one clean up suggestion.
Approving, but double check with @etam whether he has any more comments.

File: golem/network/transport/message.py (hunk 1 above)
Comment (mfranciszkiewicz) [reply golem/network/transport/message.py]: I've added `TYPE = None` to make the linter happy, but this part will fail (0 and up are reserved, negative value is not really possible). Overall, it doesn't make much sense.

Comment (etam) [reply golem/network/transport/message.py]: Without TYPE it will fail anyway, so it's absolutely fine.

**Review by etam — APPROVED** [review]

File: golem/network/transport/message.py (hunk 10 above)
Comment (mfranciszkiewicz) [reply golem/network/transport/message.py]: The difference is that `decrypt_and_deserialize` was only used in tests, so I did that freely. I suggest leaving this for the DataBuffer-related PR.

**Comment by mfranciszkiewicz** [comment]
@Krigpl @etam I've removed the `DataBuffer` dependency from message.py, in preparation for the upcoming `DataBuffer` changes.

**Review by etam — APPROVED** [review]
LGTM

**Review by Krigpl — APPROVED** [review]

**Pull request merged**